design/outlier-detection-design.md

```diff
@@ -0,0 +1,262 @@
+
+## Goals
+- Add outlier detection related configuration detection for services
+- Support configuration based on route and TCPRoute.
+- Support configuration for consecutive-5 xx and consecutive-local-origin-failure
+
+## Non-Goals
+- Not supported Outlier Detection configuration for consecutive-gateway-failure, success-rate, and failure-percentage
+
+## Detailed Design
+
+Add a new field in the CRD of httpproxy for configuring Outlier Detection of the service. The field is defined as follows:
+
+```go
+// OutlierDetection defines the configuration for outlier detection on a service.
+type OutlierDetection struct {
+    // Consecutive5xxErrors defines The number of consecutive server-side error responses before a consecutive 5xx ejection occurs.
+    // When the backend host encounters consecutive
+    // errors greater than or equal to Consecutive5xxErrors, it will be
+    // ejected from the load balancing pool.
+    // for HTTP services, a 5xx counts as an error and for TCP services
+    // connection failures and connection timeouts count as an error.
+    // It can be disabled by setting the value to 0.
+    // Defaults to 5.
+    // +optional
+    Consecutive5xxErrors *uint32 `json:"consecutive5xxErrors,omitempty"`
+    
+    // Interval is the interval at which host status is evaluated.
+    // Defaults to 10s.
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms))+)$`
+    Interval *string `json:"interval,omitempty"`
+    
+    // BaseEjectionTime is the base time that a host is ejected for.
+    // A host will remain ejected for a period of time equal to the
+    // product of the ejection base duration and the number of times the host has been ejected.
+    // Defaults to 30s.
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms))+)$`
+    BaseEjectionTime *string `json:"baseEjectionTime,omitempty"`
+    
+    // MaxEjectionTime is the maximum time a host will be ejected for.
+    // After this amount of time, a host will be returned to normal operation.
+    // If not specified, the default value (300s) or BaseEjectionTime value is applied, whatever is larger.
+    // Defaults to 300s.
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms))+)$`
+    MaxEjectionTime *string `json:"maxEjectionTime,omitempty"`
+    
+    // SplitExternalLocalOriginErrors defines whether to split the local origin errors from the external origin errors.
+    // Defaults to false.
+    // +optional
+    // +kubebuilder:default=false
+    SplitExternalLocalOriginErrors bool `json:"splitExternalLocalOriginErrors"`
```

Comment:
a nice advantage of this passive healthcheck config is we can get more granular failure details like this vs. our existing active http healthcheck configuration, I would err on the side of always enabling this to separate server errors from possible network level errors, but curious to see what others think

Reply:
+1, I would lean on the side of enabling this by default. Regarding granularity, we should also allow users to tune `consecutive_gateway_failure` and  `consecutive_local_origin_failure`; otherwise, this toggle by itself feels too rigid. 

We should also make sure to set sane defaults and that the accompanying documentation is clear.

Reply:
thanks for your review.The setting of `consecutive_local_origin_failure` is already in the design [document](https://github.com/projectcontour/contour/pull/5460/files#diff-33a757f177574a421370e2afee9143b2b85e08901c50c37ce6d34630cde7cda6R69). Consecutive_gateway_failure was [not intended to be configured at the beginning](https://github.com/projectcontour/contour/pull/5460/files#diff-33a757f177574a421370e2afee9143b2b85e08901c50c37ce6d34630cde7cda6R18). Of course, if this is a good design, I will update this design document to support `consecutive_gateway_failure`.
@clayton-gonsalves @sunjayBhatia